stablehlo/transforms/StablehloLegalizeAToB.cpp

```diff
@@ -0,0 +1,138 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"  // Include for TypeConverter
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOLEGALIZEATOBPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+// move this to common quant_util file
+bool isAnyQuantizedTypes(TypeRange types) {
+  return llvm::any_of(types, [](Type type) {
+    return isa<quant::QuantizedType>(getElementTypeOrSelf(type));
+  });
+}
+
+struct QuantizedStablehloAToBOpConversion
+    : public OpRewritePattern<stablehlo::UniformQuantizeOp> {
+  using OpRewritePattern<stablehlo::UniformQuantizeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(stablehlo::UniformQuantizeOp op,
+                                PatternRewriter& rewriter) const override {
+    auto* quantizingOp = op->getOperand(0).getDefiningOp();
+    // Ignore UniformQuantized op withput the parent op
+    if (!(quantizingOp)) return failure();
+
+    // if quantizingOp is alredy quanitzed, no work to do
+    if (isAnyQuantizedTypes(quantizingOp->getOperandTypes())) return failure();
+
+    llvm::SmallVector<Value> quantizedOperands;
+    for (const auto& operandx : quantizingOp->getOperands()) {
+      // backtrack and get the quantized operand
+      // quantizedOp -> DequantizedOp -> shloOp
+      auto* opx = operandx.getDefiningOp();
+      // It is possible an operand is input of the module.
+      if (!(opx)) {
+        quantizedOperands.push_back(operandx);
+        continue;
+      }
+
+      if (auto deqOp = dyn_cast<stablehlo::UniformDequantizeOp>(opx)) {
+        quantizedOperands.push_back(deqOp->getOperand(0));
+      } else {
+        // can argument is result of another quantized op and not
+        // UniformDequantizedOp? ignore this use case for now.
+        return failure();
+      }
+    }
+
+    SmallVector<Type> quantizedResults;
+    for (auto newResult : op->getResultTypes()) {
+      quantizedResults.push_back(newResult);
+    }
+    rewriter.setInsertionPointAfter(quantizingOp);
+    OperationState newState(
+        quantizingOp->getLoc(), quantizingOp->getName().getStringRef(),
+        quantizedOperands, quantizedResults, quantizingOp->getAttrs());
+    for (unsigned int i = 0; i < quantizingOp->getNumRegions(); ++i) {
+      newState.addRegion();
+    }
+    Operation* quantizedOp = rewriter.create(newState);
```

Comment:
Can we modify the existing op in place somehow? Not sure if that will work, since I'm not sure if we can change result type.

Reply:
Abhinav: Not sure if we can manipulate the result type.

Reply:
>>  modify the existing op in place

Just to add removing the original fp op might not be desirable when 
 
```
dq = stablehlo.uniform.dequantize
op = stablehlo.abs(dq)
q = stablehlo.uniform.quantize(op)
non-q = stablehlo.add(op) // another use of op in non-quantized context
```
where we might need to preserve the original fp op. The other option of of entirely removing the fp op and instead  dequantizing the quantized op to be fed to `non-q` will not have the same semantics  mainly because of "loss-of-precision" reasons.

Reply:
created https://github.com/openxla/stablehlo/issues/2485 to track this as further enhancements to the pass.